This commit is to fix the way that files are being referred to in the static handler

adiraju13 opened the pull request:
@RyanLPeterman  check it out.These changes are to make sure that we are referring to the files with the correct path. Let me know if you have any questions. Also found the reason for the bug we were having that was crashing our server with static files.

Commit:
This commit is to fix the way that files are being referred to in the static handler

Commit:
This commit corrects get_longest_prefix, and makes it clean and easy to understand.

adiraju13 commented:
I also just added the longest prefix matching, check it out. its pretty clean IMO, a better solution than matching by "/...../", which was just too hard to understand.

ryanlpeterman commented on handler.cpp:
```diff
@@ -64,17 +64,41 @@ RequestHandler::Status StaticHandler::Init(const std::string& uri_prefix, const
+    
+    std::string abs_path = serve_path;
+    //get the file name/path from the request url
+    std::string file_path = request.uri().substr(uri_prefix.length());
```
love it

ryanlpeterman commented on handler.cpp:
```diff
@@ -64,17 +64,41 @@ RequestHandler::Status StaticHandler::Init(const std::string& uri_prefix, const
+    //get the file name/path from the request url
+    std::string file_path = request.uri().substr(uri_prefix.length());
+
+    //Code to add the / between the file path and serve_path properly
```
what do you think about moving this bit that preprocesses our serve_path to in the init file? This computation should only need to happen once and after that we use the same serve_path every time right?

adiraju13 replied:
what this does is it gets the file part of the request, not the prefix part. 

For example, if the user asks for /static/foo.txt, and our prefix is /static, then file_path will equal /foo.txt. I think this needs to be done for each request.

adiraju13 replied:
If you want I can write a comment to make it more clear in the code.

adiraju13 replied:
Nevermind I see what you are saying, let me make the fixes now.

ryanlpeterman commented on handler.cpp:
```diff
@@ -64,17 +64,41 @@ RequestHandler::Status StaticHandler::Init(const std::string& uri_prefix, const
 
+    std::cout << "Attempting to serve file from: " << serve_path << std::endl;
+
+    //WARNING WE SHOULD LOOK AT THIS. MIGHT CAUSE AN ERROR BECAUSE
```
I believe I have a solution to this:
http://stackoverflow.com/questions/328944/how-do-i-check-if-a-file-is-a-regular-file-in-c

I'm down to merge this them implement it, just a small change in the code of file_exists. I remember when I was implementing it boost::filesystem was giving me problems but it seems to be working for everyone else. Looks simple enough.

adiraju13 replied:
yeee sounds good

ryanlpeterman commented on session.cpp:
```diff
@@ -105,32 +105,15 @@ std::string session::resetUri(const std::string original_uri, const std::string
-      if (handler_map.count("/") == 0){
-        break;
+  unsigned int longest_match_size = 0;
+  std::string longest_match = "";
```
this is so clean jesus

adiraju13 replied:
you flatter me 😊

ryanlpeterman commented on session.cpp:
```diff
@@ -39,11 +39,10 @@ int session::handle_request(const boost::system::error_code& error,
+  // find longest prefix that uniquely matches uri (our key for handler_map)
+  std::string longest_prefix = get_longest_prefix(request->uri());
+  // get corresponding handler
+  std::shared_ptr<RequestHandler> handler_ptr = handler_map[longest_prefix];
```
we should add a handler in this map of the following:

"" -> default handler

Since get_longest_prefix returns empty string if it finds no match

adiraju13 replied:
Im super down for that. that makes sense to me.

Commit:
Changes to make the serve_path be set once instead of for every request.
Also changed the simple config file to have the static root refer to our current directory.

Commit:
Some typo fixes to setting the serve_path once

ryanlpeterman commented on handler.cpp:
```diff
@@ -65,16 +65,40 @@ RequestHandler::Status StaticHandler::Init(const std::string& uri_prefix, const
+
+    //Code to add the / between the file path and serve_path properly
+    if (serve_path.length() != 0 && serve_path[serve_path.length()-1] != '/'){
+        if (uri_prefix.length() != 0 && uri_prefix[uri_prefix.length()-1] == '/'){
```
hell yea

ryanlpeterman commented on handler.cpp:
```diff
@@ -65,16 +65,40 @@ RequestHandler::Status StaticHandler::Init(const std::string& uri_prefix, const
 
+    std::cout << "Attempting to serve file from: " << serve_path << std::endl;
+
+    // TODO WE SHOULD LOOK AT THIS. MIGHT CAUSE AN ERROR BECAUSE
```
ty for TODO
